src/main/kotlin/org/pkl/lsp/ast/PklNode.kt

```diff
@@ -741,112 +708,175 @@ abstract class AbstractPklNode(
   }
 }
 
-fun List<ParseTree>.toNode(project: Project, parent: PklNode?, idx: Int): PklNode? {
+class PklErrorImpl(
+  override val project: Project,
+  override val parent: PklNode?,
+  override val ctx: TreeSitterNode,
+) : PklError, AbstractPklNode(project, parent, ctx) {
+  override fun <R> accept(visitor: PklVisitor<R>): R? {
+    return visitor.visitError(this)
+  }
+}
+
+fun List<TreeSitterNode>.toNode(project: Project, parent: PklNode?, idx: Int): PklNode? {
   return get(idx).toNode(project, parent)
 }
 
-fun ParseTree.toNode(project: Project, parent: PklNode?): PklNode? {
-  return when (this) {
+private fun TreeSitterNode.toTypeNode(project: Project, parent: PklNode?): PklNode? {
+  return when {
+    text == "unknown" -> PklUnknownTypeImpl(project, parent!!, this)
+    text == "nothing" -> PklNothingTypeImpl(project, parent!!, this)
+    text == "module" -> PklModuleTypeImpl(project, parent!!, this)
+    childCount <= 0 -> null
+    children[0].type == "stringConstant" -> PklStringLiteralTypeImpl(project, parent!!, this)
+    children[0].type == "qualifiedIdentifier" -> PklDeclaredTypeImpl(project, parent!!, this)
+    children[0].type == "(" && childCount == 3 -> PklParenthesizedTypeImpl(project, parent!!, this)
+    children[0].type == "(" -> PklFunctionTypeImpl(project, parent!!, this)
+    // TODO: for pkl-tree-sitter `*Foo` alone is a valid type
+    children[0].type == "*" -> PklDefaultUnionTypeImpl(project, parent!!, this)
```

Comment:
I think this is fine; we can add a diagnostic to flag this as an error through the language server. It might be even preferable to do it this way because we can provide a better error message and a quickfix.